### ISSUE TITLE
Support OpenCL.jl as device runtime

### ISSUE THREAD

**Issue by jpsamaroo** [JuliaGPU/AMDGPUnative.jl#20]
Currently, we only support using HSARuntime.jl as our device runtime, which currently only allows use of this package by Linux users with recent kernels. However, @vchuravy has made the suggestion that we could use OpenCL.jl as an alternative backend, since it supports loading arbitrarily-generated GPU object code. Because AMD OpenCL drivers are available on Windows and Mac systems (and potentially others, like the BSDs), we should be able to use this as a workaround for users on such systems until ROCm is made available.

**Comment by vchuravy** [comment]
Yeah this should be relatively straight-forward the tricky bit is to get the calling convention right. 
https://llvm.org/docs/AMDGPUUsage.html#opencl there are a couple of implicit arguments we need to expose.
